File: ui/src/gql/dailyTracker.ts
import {gql} from 'apollo-boost';

export const DailyTrackers = gql`
    query DailyTrackers {
        dailyTrackers {
            id
            name
            type
            chainable
            color
            streakDays
            milestoneBadges
        }
    }
`;

export const DailyTrackerValues = gql`
    query DailyTrackerValues($from: String!, $to: String!) {
        dailyTrackerValues(fromInclusive: $from, toInclusive: $to) {
            id
            trackerId
            tracker {
                id
                name
                type
                color
                chainable
            }
            date
            textValue
            numberValue
            boolValue
        }
    }
`;

export const CreateDailyTracker = gql`
    mutation CreateDailyTracker($name: String!, $type: TrackerType!, $chainable: Boolean!, $color: String!) {
        createDailyTracker(name: $name, type: $type, chainable: $chainable, color: $color) {
            id
            name
            type
            chainable
            color
            streakDays
            milestoneBadges
        }
    }
`;

export const UpdateDailyTracker = gql`
    mutation UpdateDailyTracker($id: Int!, $name: String, $type: TrackerType, $chainable: Boolean, $color: String) {
        updateDailyTracker(id: $id, name: $name, type: $type, chainable: $chainable, color: $color) {
            id
            name
            type
            chainable
            color
            streakDays
            milestoneBadges
        }
    }
`;

export const RemoveDailyTracker = gql`
    mutation RemoveDailyTracker($id: Int!) {
        removeDailyTracker(id: $id) {
            id
        }
    }
`;

export const SetDailyTrackerValue = gql`
    mutation SetDailyTrackerValue($trackerId: Int!, $date: String!, $textValue: String, $numberValue: Float, $boolValue: Boolean) {
        setDailyTrackerValue(trackerId: $trackerId, date: $date, textValue: $textValue, numberValue: $numberValue, boolValue: $boolValue) {
            id
            trackerId
            tracker {
                id
                name
                type
                color
                chainable
            }
            date
            textValue
            numberValue
            boolValue
        }
    }
`;

export const RemoveDailyTrackerValue = gql`
    mutation RemoveDailyTrackerValue($id: Int!) {
        removeDailyTrackerValue(id: $id) {
            id
        }
    }
`;

File: model/dailytracker.go
package model

import (
	"database/sql/driver"
	"fmt"
)

// DailyTracker defines a daily tracker with type-safe value storage
type DailyTracker struct {
	ID        int        `gorm:"primary_key;unique_index;AUTO_INCREMENT"`
	UserID    int        `gorm:"type:int REFERENCES users(id) ON DELETE CASCADE"`
	Name      string
	Type      TrackerType
	Chainable bool // Enables streak tracking
	Color     string
}

// TrackerType represents the value type (text/number/boolean)
type TrackerType string

// Value for db
func (t TrackerType) Value() (driver.Value, error) {
	return string(t), nil
}

// Scan for db
func (t *TrackerType) Scan(value interface{}) error {
	s, ok := value.([]byte)
	if !ok {
		return fmt.Errorf("expected string but was %#v", value)
	}
	*t = TrackerType(s)
	return nil
}

// TrackerType constants
const (
	TypeText     TrackerType = "text"
	TypeNumber   TrackerType = "number"
	TypeBoolean  TrackerType = "boolean"
)

// DailyTrackerValue stores a single day's tracker entry with type-safe columns
type DailyTrackerValue struct {
	ID          int    `gorm:"primary_key;unique_index;AUTO_INCREMENT"`
	TrackerID   int    `gorm:"type:int REFERENCES daily_trackers(id) ON DELETE CASCADE"`
	Date        string // Format: YYYY-MM-DD (user's timezone date)

	// Type-safe value columns (only one is used based on tracker type)
	TextValue   string  `gorm:"type:text"`
	NumberValue float64 `gorm:"type:float"`
	BoolValue   bool    `gorm:"type:boolean"`

	CreatedAt   Time
	UpdatedAt   Time

	// Relationship
	Tracker     DailyTracker `gorm:"foreignkey:TrackerID"`
}
